src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/CompareSymbolsCorrectlyFix.cs

```diff
@@ -65,25 +68,68 @@ private static async Task<Document> EnsureEqualsCorrectAsync(Document document,
             var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
             var generator = editor.Generator;
 
+            // With "a?.b?.c?.Equals(b)", the invocation operation syntax is only ".Equals(b)".
+            // Walk-up the tree to find all parts of the conditional access and also the root
+            // so that we can replace the whole expression.
+            var conditionalAccessMembers = new List<SyntaxNode>();
+            IOperation currentOperation = invocationOperation;
+            while (currentOperation.Parent is IConditionalAccessOperation conditionalAccess)
+            {
+                currentOperation = conditionalAccess;
+                conditionalAccessMembers.Add(conditionalAccess.Operation.Syntax);
+            }
+
+            var arguments = GetNewInvocationArguments(invocationOperation, conditionalAccessMembers);
+            var replacement = generator.InvocationExpression(GetEqualityComparerDefaultEquals(generator), arguments);
+
+            var nodeToReplace = currentOperation.Syntax;
+            editor.ReplaceNode(nodeToReplace, replacement.WithTriviaFrom(nodeToReplace));
+
+            return editor.GetChangedDocument();
+        }
+
+        private IEnumerable<SyntaxNode> GetNewInvocationArguments(IInvocationOperation invocationOperation,
+            List<SyntaxNode> conditionalAccessMembers)
+        {
             var arguments = invocationOperation.Arguments.Select(argument => argument.Syntax);
 
-            if (invocationOperation.Instance is not null)
+            if (invocationOperation.GetInstance() is not IOperation instance)
             {
-                arguments = new[] { invocationOperation.Instance.Syntax }.Concat(arguments);
+                return arguments;
             }
 
-            var replacement = generator.InvocationExpression(
-                                    GetEqualityComparerDefaultEquals(generator),
-                                    arguments);
+            if (instance.Kind != OperationKind.ConditionalAccessInstance)
+            {
+                return new[] { instance.Syntax }.Concat(arguments);
+            }
 
-            editor.ReplaceNode(invocationOperation.Syntax, replacement.WithTriviaFrom(invocationOperation.Syntax));
+            // We need to rebuild a new conditional access chain skipping the invocation
+            if (conditionalAccessMembers.Count == 0)
+            {
+                throw new InvalidOperationException("Invocation contains conditional expression but we could not detect the parts.");
+            }
+            else if (conditionalAccessMembers.Count == 1)
+            {
+                return new[] { conditionalAccessMembers[0] }.Concat(arguments);
+            }
+            else
+            {
+                var currentExpression = conditionalAccessMembers.Count > 2
```

Comment:
Do you want to file a follow-up issue to simplify this code and directly use Roslyn public API once we move to the new Microsoft.CodeAnalysis in future? Actually, having said that, I think it is about time to move some of analyzer packages to newer Microsoft.CodeAnalysis API, I can do that next week.

Reply:
https://github.com/dotnet/roslyn-analyzers/issues/4580